### ISSUE TITLE
How to prepare datasets/CIHP/labels and datasets/CIHP/edges

### ISSUE THREAD

**Issue by rcrvano** [Engineering-Course/CIHP_PGN#26]
Hello!

Can you explain please how to prepare files, which should be located in this directories:
**datasets/CIHP/labels** and **datasets/CIHP/edges**

I've prepared the image file datasets/CIHP/images/image.jpg,

file datasets/CIHP/list/val_id.txt with content:
```
image
```

and file datasets/CIHP/list/val.txt with content:
```
images/image.jpg /labels/image.png
```

But I don't understand how I can generate datasets/CIHP/labels/image.png 
and datasets/CIHP/edges/image.png  ?

Because I need them for test_pgn.py.
You answered [here](https://github.com/Engineering-Course/CIHP_PGN/issues/21) that in this dir can be placed any png image. But if I just convert image.jpg to image.png and put converted file into labels and edges dir the execution of the **test_pgn.py** failed with error

```
InvalidArgumentError (see above for traceback): assertion failed: [`labels` out of bound] [Condition x < y did not hold element-wise:] [x (mean_iou/confusion_matrix/control_dependency:0) = ] [255 255 255...] [y (mean_iou/ToInt64_1:0) = ] [20]
```

Please explain how to prepare image files for starting the segmentation process?
Because your README.md file doesn't explain how to do it. Just "Prepare the images and store in $HOME/datasets". But what steps need to be done to prepare images in datasets dir?

> Inference
> - Download the pre-trained model and store in $HOME/checkpoint.
> - Prepare the images and store in $HOME/datasets.
> - Run test_pgn.py.
> - The results are saved in $HOME/output
> - Evaluation scripts are in $HOME/evaluation. Copy the groundtruth files (in Instance_ids folder) - into $HOME/evaluation/Instance_part_val before you run the script.

**Comment by Engineering-Course** [comment]
Yes. You can refer to our example images.  Or just download the edge images and change the name.

**Comment by rcrvano** [comment]
Yes. I refer to the example.

The input directory has the following structure:
```
./datasets/CIHP/list
./datasets/CIHP/list/val_id.txt
./datasets/CIHP/list/train.txt
./datasets/CIHP/list/val.txt
./datasets/CIHP/list/train_id.txt
./datasets/CIHP/list/train_rev.txt
./datasets/CIHP/labels
./datasets/CIHP/labels/0032190.png
./datasets/CIHP/images
./datasets/CIHP/images/0032190.jpg
./datasets/CIHP/edges
./datasets/CIHP/edges/0032190.png
```

This file ./datasets/CIHP/images/0032190.jpg - it is an target image file.
![0032190](https://user-images.githubusercontent.com/19143863/56952327-adead880-6b6c-11e9-875b-0e8951844da4.jpg)

And as you can see this file ./datasets/CIHP/labels/0032190.png already partially segmented.
![0032190](https://user-images.githubusercontent.com/19143863/56952361-c65af300-6b6c-11e9-897b-b57ea02fd58d.png)

This file: ./datasets/CIHP/edges/0032190.png seems just dark image (but I'm not sure)
![0032190](https://user-images.githubusercontent.com/19143863/56952586-37020f80-6b6d-11e9-8d81-e367cc27889b.png)

Can you tell me please. How I can automatically generate this file **labels/0032190.png** from initial target image file **images/0032190.jpg**? Because if I replace *images/0032190.jpg**  with my own, the test script test_pgn.py doesn't want to work, beacuse label image file is wrong. 

Please, explain me how I can automatically generate labels input files for CIHP_PGN?

**Comment by Engineering-Course** [comment]
labels/*.png should have only one channel. Have a try?

**Comment by rcrvano** [comment]
If I create a grayscale png image from source image I receive this error anyway 
```
tensorflow.python.framework.errors_impl.InvalidArgumentError: assertion failed: [`labels` out of bound] [Condition x < y did not hold element-wise:] [x (mean_iou/confusion_matrix/control_dependency:0) = ] [252 252 252...] [y (mean_iou/ToInt64_1:0) = ] [20]
```

There is another image from your example: 

Target image 
![0026375](https://user-images.githubusercontent.com/19143863/57215131-97d29180-701e-11e9-8dd6-da5cb99a300b.jpg)

And label image
![0026375](https://user-images.githubusercontent.com/19143863/57215114-8ee1c000-701e-11e9-9cd7-b0ba14323cf7.png)

I don't understand how to prepare an input data (labels and edges) for your project?

**Comment by LogWell** [comment]
感觉作者的回复有些敷衍, 希望你能把README写得更清楚点~

**Comment by rcrvano** [comment]
Your README file includes following instruction

> **Inference**
> 1) Download the pre-trained model and store in $HOME/checkpoint.
> 2) Prepare the images and store in $HOME/datasets.
> 3) Run test_pgn.py.

But how to prepare images for datasets??

**Comment by rcrvano** [comment]
@LogWell found a solution how to prepare DATASETS. You can use following files structure:

```
datasets/CIHP/images/0002190.png
datasets/CIHP/list/img_list.txt
datasets/CIHP/images/tool/logwell_script.m
datasets/CIHP/images/edges
datasets/CIHP/images/labels
```
where 
datasets/CIHP/images/0002190.png - it's a source image

datasets/CIHP/list/img_list.txt contain following data:
`0002190.png`

and  datasets/CIHP/images/tool/logwell_script.m - it's a MATLAB script for prepare edges and labels:
```
clear;
close all;
fclose all;
%%
imglist = '../list/img_list.txt';  % 00000.png
list = textread(imglist, '%s');

for i = 1:length(list);
    imname = list{i};
    instance_map = imread(fullfile('../images', imname));
    instance_contour = uint8(imgradient(rgb2gray(instance_map)) > 0);
    imwrite(instance_contour, fullfile('../edges', imname));
    imwrite(instance_contour, fullfile('../labels', imname));
    
end

```


Just run this script using command like:
`/opt/MATLAB/R2018b/bin/matlab -nodisplay -nojvm -nosplash -nodesktop -r "try, run('tool/logwell_script.m'), catch, exit(1), end, exit(0);"`

and after that you can run 
```python test_pgn.py```
for get segmented images.

**Comment by MISSXUAN0** [comment]
@rcrvano 
Hi,thank you for your solution for prepare DATASETS. Just put the data in the specified path,then test_pgn.py run success？
I missed this problem：
 (octopus) test@True3d-pc2:~/octopus/CIHP_PGN-master$ python test_pgn.py
WARNING:tensorflow:From /home/test/anaconda3/envs/octopus/lib/python2.7/site-packages/tensorflow/python/training/input.py:187: __init__ (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
WARNING:tensorflow:From /home/test/anaconda3/envs/octopus/lib/python2.7/site-packages/tensorflow/python/training/input.py:187: add_queue_runner (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
WARNING:tensorflow:From test_pgn.py:39: calling expand_dims (from tensorflow.python.ops.array_ops) with dim is deprecated and will be removed in a future version.
Instructions for updating:
Use the `axis` argument instead
Segmentation fault (core dumped)

**Comment by MISSXUAN0** [comment]
Thank you for the your method. This method can get the result when testing the picture of CIHP, but can't get result on my own prepared data.
 Can you get the results directly？Just run python test_pgn.py and don‘t need to train on own dataset？

**Comment by rcrvano** [comment]
Yes.  Right. Prepare datasets as described above and then just run test_pgn

**Comment by MISSXUAN0** [comment]
ok.Thank you.I got the result.
The reason why it has been unsuccessful is that the picture pixels are too large.

**Comment by sepidemam** [comment]
> ok.Thank you.I got the result.
> The reason why it has been unsuccessful is that the picture pixels are too large.

I get the same error. Can you please let me know how did you fix pixel issue?

**Comment by DomhnallBoyle** [comment]
Hi @Engineering-Course 

I'm trying to repeat the preparation of images for segmentation using the following image from the datasets folder: 
![0002190](https://user-images.githubusercontent.com/20018229/60662058-447c9400-9e53-11e9-9e77-6a22679e5fd3.jpg)

I'm using the matlab script located [here]() with some of the changes mentioned by @rcrvano above. I'm getting the following results for the edges: 

![0002190](https://user-images.githubusercontent.com/20018229/60662286-dbe1e700-9e53-11e9-8451-07936265aaac.png)

This is the output of the same image already located in the edges folder: 
![image](https://user-images.githubusercontent.com/20018229/60662264-c967ad80-9e53-11e9-9513-ac3870a175e6.png)

Why can I not reproduce the edges in the same way as the image above?

**Comment by neonb88** [comment]
I think there's confusion about this.  You don't actually have to install MATLAB or Octave, create good `edges` and `labels`.  That won't make the result better.  The only reason it takes in edges and labels at all is to measure accuracy and precision.

I just wrote a method that doesn't require MATLAB or Octave, only the `python` modules PGN already requires:

```
import numpy as np
import imageio as ii
from PIL import Image
nxb_img   = Image.open('datasets/CIHP/images/front.jpg')      # This is your image.
 
# Reshape their label image to our size 
label_img = Image.open('datasets/CIHP/labels/0005008.png')  # This is the label image from CIHP.
nxb_label_img = label_img.resize(nxb_img.size, Image.NEAREST)
nxb_label_img.save('datasets/CIHP/labels/front.png')

# Reshape their edge image to our size 
edge_img  = Image.open('/home/nathanbendich/CIHP_PGN/datasets/CIHP/edges/0005008.png')
nxb_edge_img  = edge_img.resize(nxb_img.size, Image.NEAREST)
nxb_edge_img.save('datasets/CIHP/edges/front.png')
```

You can use Logwell's MATLAB code, but it won't improve performance.
